src/internal/Subscription.ts

```diff
@@ -53,180 +61,188 @@ export class Subscription implements SubscriptionLike {
   add(teardown: TeardownLogic): void {
-    let subscription = (<Subscription>teardown);
-
-    if (!teardown) {
-      return;
-    }
-
-    switch (typeof teardown) {
-      case 'function':
-        subscription = new Subscription(<(() => void)>teardown);
-      case 'object':
-        if (subscription === this || subscription.closed || typeof subscription.unsubscribe !== 'function') {
-          // This also covers the case where `subscription` is `Subscription.EMPTY`, which is always in `closed` state.
-          return;
-        } else if (this.closed) {
-          subscription.unsubscribe();
-        } else if (!(subscription instanceof Subscription)) {
-          const tmp = subscription;
-          subscription = new Subscription();
-          subscription._subscriptions = [tmp];
+    // Only add the teardown if it's not undefined
+    // and don't add a subscription to itself.
+    if (teardown && teardown !== this) {
+      if (this.closed) {
+        // If this subscription is already closed,
+        // execute whatever teardown is handed to it automatically.
+        if (typeof teardown === 'function') {
+          teardown();
+        } else {
+          teardown.unsubscribe();
         }
-        break;
-      default: {
-        throw new Error('unrecognized teardown ' + teardown + ' added to Subscription.');
+      } else {
+        this._teardowns = this._teardowns ?? [];
+        if (teardown instanceof Subscription) {
+          // We don't add closed subscriptions, and we don't add the same subscription
+          // twice. Subscription unsubscribe is idempotent.
+          if (teardown.closed || teardown._hasParent(this)) {
+            return;
+          }
+          teardown._addParent(this);
+        }
+        this._teardowns.push(teardown);
       }
     }
+  }
+
+  private _hasParent(parent: Subscription) {
+    return this._singleParent === parent || this._parents?.includes(parent) || false;
+  }
 
-    // Add `this` as parent of `subscription` if that's not already the case.
-    let { _parentOrParents } = subscription;
-    if (_parentOrParents == null) {
-      // If we don't have a parent, then set `subscription._parents` to
-      // the `this`, which is the common case that we optimize for.
-      subscription._parentOrParents = this;
-    } else if (_parentOrParents instanceof Subscription) {
-      if (_parentOrParents === this) {
-        // The `subscription` already has `this` as a parent.
-        return;
+  private _addParent(parent: Subscription) {
+    // If we already have one or more parents, we need
+    // to add to the list of parents.
+    if ((this._parents || this._singleParent) && this._singleParent !== parent) {
+      if (this._singleParent) {
+        // We are adding our second parent.
+        // Move the single parent over to the parents list.
+        this._parents = [this._singleParent];
+        this._singleParent = undefined;
+      }
+      if (!this._parents!.includes(parent)) {
```

Comment:
issue: There is no need to check this here, AFAICT. `_hasParent` is called before `_addParent` is called - it's called in only one place - so this check means that the perf penalty of scanning the array is paid twice.